### PR TITLE
Refactor entry point IR

### PR DESCRIPTION
**Pull request by kvark**
Another innocent change that turned out to be rather big. I actually was just trying to update WGSL, and the changes started being contagious...

On the IR side, this changes entry points to:
  - be a map, so no collisions by construction
  - store the function directly instead of a handle, so that nothing can call this function

It also renames the "set" to "group" in the decorations.

And it updates WGSL attribute syntax to the one that just got accepted.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -718,12 +714,12 @@ pub struct Function {
...
-    pub stage: ShaderStage,
-    /// Name identifying this entry point.
-    pub name: String,
-    /// The function to be used.
-    pub function: Handle<Function>,
+    /// Early depth test for fragment stages.
+    pub early_depth_test: Option<EarlyDepthTest>,
+    /// Workgroup size for compute stages
+    pub workgroup_size: [u32; 3],
+    /// The function body.
```
Comment (Napokue) [inline src/lib.rs:721]: Minor, but maybe crucial for people new to the codebase: this isn't actually only the function body, but also the signature and all other details associated with a function. I think this comment is more confusing than having no comment.

File: src/lib.rs [hunk 2]
```diff
@@ -416,8 +412,8 @@ pub enum Binding {
     BuiltIn(BuiltIn),
     /// Indexed location.
     Location(u32),
-    /// Binding within a descriptor set.
-    Descriptor { set: u32, binding: u32 },
+    /// Binding within a resource group.
+    Resource { group: u32, binding: u32 },
```
Comment (Napokue) [inline src/lib.rs:416]: I'm just curious, but why did you rename it to `group` instead of `set`?

File: src/front/wgsl/mod.rs [hunk 3]
```diff
@@ -1373,30 +1361,56 @@ impl Parser {
         let mut binding = None;
         // Perspective is the default qualifier.
         let mut interpolation = None;
+        let mut stage = None;
```
Comment (Napokue) [inline src/front/wgsl/mod.rs:1364]: Maybe our code base isn't there yet with syntax validation, but what happens if an user forgets to add the `stage` attribute? I think it will completely go through the front-end -> IR, and crashes at the back-end?

I see at line [1509](https://github.com/gfx-rs/naga/pull/200/files#diff-37ba88792a802ec39755d9da39194f86R1509) that it then gets interpreted as a normal function.

File: src/front/glsl/parser.rs [hunk 4]
```diff
@@ -828,9 +828,19 @@ pomelo! {
...
-        let name = f.name.clone();
-        let handle = extra.functions.append(f);
-        if let Some(name) = name {
+        if f.name == extra.entry {
+            let name = extra.entry.take().unwrap();
+            extra.module.entry_points.insert(
+                (extra.shader_stage, name),
+                EntryPoint {
+                    early_depth_test: None,
+                    workgroup_size: [1; 3], //TODO
```
Comment (Napokue) [inline src/front/glsl/parser.rs:837]: Do you still need to do something here, or is it for a later?

File: src/back/msl.rs [hunk 5]
```diff
@@ -924,8 +907,56 @@ impl<W: Write> Writer<W> {
...
+
+            for (index, &ty) in fun.parameter_types.iter().enumerate() {
+                let name = Name::from(ParameterIndex(index));
+                let member_type_name = module.types[ty].name.or_index(ty);
+                let separator = separate(index + 1 == fun.parameter_types.len());
+                writeln!(self.out, "\t{} {}{}", member_type_name, name, separator)?;
+            }
+            writeln!(self.out, ") {{")?;
+
+            // write down function body
```
Comment (Napokue) [inline src/back/msl.rs:924]: Nit: I think this is a leftover comment from down below, as I have seen it at line [1137](https://github.com/gfx-rs/naga/pull/200/files#diff-aaa1f30b3e995f65d9bd7557d98981fdR1137).

File: src/back/msl.rs [hunk 6]
```diff
@@ -941,206 +972,175 @@ impl<W: Write> Writer<W> {
...
-                }
-            }
-            let output_name = fun.name.or_index(OutputStructIndex(fun_handle));
+
+            let output_name = Name {
+                class: "Output",
+                source: NameSource::Custom { name, prefix: true },
+            };
 
             // make dedicated input/output structs
```
Comment (Napokue) [inline src/back/msl.rs:981]: Nit: Leftover comment?

**Review by Napokue — CHANGES_REQUESTED** [review]
Thank you, I love the changes, especially those new Stage attributes!

File: test-data/simple/simple.wgsl [hunk 7]
```diff
@@ -1,10 +1,10 @@
 # vertex
-[[location 0]] var<in> a_pos : vec2<f32>;
-[[location 0]] var<out> o_pos : vec4<f32>;
+[[location(0)]] var<in> a_pos : vec2<f32>;
+[[location(0)]] var<out> o_pos : vec4<f32>;
 
+[[stage(vertex)]]
```
[the remark this answers is not in the document]
Comment (kvark) [reply test-data/simple/simple.wgsl:5]: I proposed it :)

File: test-data/simple/simple.expected.ron [hunk 8]
```diff
@@ -64,58 +64,56 @@
...
+                    ),
+                    Return(
+                        value: None,
+                    ),
+                ],
+            ),
         ),
-    ],
+    },
 )
```
[the remark this answers is not in the document]
Comment (kvark) [reply test-data/simple/simple.expected.ron:119]: seems like it's passing, so no?

File: src/front/wgsl/mod.rs (hunk 3 above)
Comment (kvark) [reply src/front/wgsl/mod.rs:1364]: If there is no `stage`, then it's not an entry point. IR will be built just fine. If we are talking about wgpu-core as the client, it will try to get the entry point information from IR, and it will not find the entry point, so this is where an error will be produced.

File: src/front/glsl/parser.rs (hunk 4 above)
Comment (kvark) [reply src/front/glsl/parser.rs:837]: that matches the current behavior, so there is no need to fix this in this PR. I wanted to leave a note here to address later by somebody.

File: src/back/msl.rs (hunk 6 above)
Comment (kvark) [reply src/back/msl.rs:981]: moved it to a better place

File: test-data/simple/simple.expected.ron (hunk 8 above)
[the remark this answers is not in the document]
Comment (kvark) [reply test-data/simple/simple.expected.ron:119]: I'll add a newline
